### PR TITLE
Ensure that we only depend on a small expected set of external symbols in the freestanding stdlib

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

**Review by compnerd — CHANGES_REQUESTED** [review]
I like the idea here, but I think that this is better implemented as a python script.  That should also allow for this to extend into other OSes.  There are definitely cases of symbols which are Darwin specific here.  We do have some tests that run python scripts - that would also make the test a bit less reliant on the host tools.

**Comment by kubamracek** [comment]
👍 All sounds good, will tweak the PR.

Yes, the current list is for Darwin. I actually have a "plan" to keep shrinking this list down so that all Darwin APIs are completely gone (with the exception of one dyld API to enumerate loaded modules, see the discussion we had in the other PR).

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: utils/check_freestanding_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    "__NSGetArgc", "__NSGetArgv", "___stack_chk_fail", "___stack_chk_guard",
+    "_asl_log",
+]
+################################################################################
+
+
+if args.vendor == "apple":
+    vendor_specific_dependencies = vendor_apple_specific_dependencies
+
+    lines = subprocess.check_output(["nm", "--undefined-only", args.library]) \
```
Comment (compnerd) [inline utils/check_freestanding_dependencies.py:70]: Would be nice to have this be a parameter so we can pass it `%llvm-nm`.

**Review by compnerd — APPROVED** [review]

File: utils/check_freestanding_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+allowlist = set(common_expected_dependencies + vendor_specific_dependencies)
+for symbol in deps:
+    if symbol not in allowlist:
+        print("Error: Unexpected dependency '{}'".format(symbol))
+        fail = True
+
+for symbol in allowlist:
+    if symbol not in deps:
+        print("Warning: Allowed dependency '{}' not present".format(symbol))
+        fail = True
```
Comment (kubamracek) [inline utils/check_freestanding_dependencies.py:116]: Note to self, this shouldn't set fail = True. It should be just a warning.

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by gottesmm** [comment]
I am for this and think it is reasonable/important to make sure mistakes to sneak in. The only question I have is if this is a BuildSystem unit test or a stdlib unit test. I am fine in either place though.

**Comment by kubamracek** [comment]
Ah, forgot to explain: The reason I put it into test/stdlib/ is to make it actually run in the "minimal" buildbot we have on ci.swift.org. Because that bot only builds the runtime+stdlib and not the compiler, it only runs tests from a subset of lit tests.

If you're okay with that, then I'd like to keep the test in test/stdlib/, and perhaps figure our if there other sub-folders of tests that we should be running later.

**Comment by gottesmm** [comment]
@kubamracek SGTM. Can you put in a big comment explaining that? I just don't want someone to move it by mistake.

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Pull request merged**